taichi/analysis/gather_uniquely_accessed_pointers.cpp

```diff
@@ -0,0 +1,157 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/analysis.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/visitors.h"
+#include <algorithm>
+
+TLANG_NAMESPACE_BEGIN
+
+class LoopUniqueStmtSearcher : public BasicStmtVisitor {
+ private:
+  std::unordered_set<Stmt *> loop_invariant_;
+
+  // If loop_unique_[stmt] is -1, the value of stmt is unique among the
+  // top-level loop.
+  // If loop_unique_[stmt] is x > 0, the value of stmt is unique to
+  // the x-th loop index.
+  std::unordered_map<Stmt *, int> loop_unique_;
+
+ public:
+  int num_different_loop_indices{-1};
+  using BasicStmtVisitor::visit;
+
+  LoopUniqueStmtSearcher() {
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  void visit(LoopIndexStmt *stmt) override {
+    if (stmt->loop->is<OffloadedStmt>())
+      loop_unique_[stmt] = stmt->index;
+  }
+
+  void visit(LoopUniqueStmt *stmt) override {
+    loop_unique_[stmt] = -1;
+  }
+
+  void visit(ConstStmt *stmt) override {
+    loop_invariant_.insert(stmt);
+  }
+
+  void visit(UnaryOpStmt *stmt) override {
+    if (loop_invariant_.count(stmt->operand) > 0) {
+      loop_invariant_.insert(stmt);
+    }
+    if (loop_unique_.count(stmt->operand) > 0 &&
+        (stmt->op_type == UnaryOpType::neg)) {
+      // TODO: Other injective unary operations
+      loop_unique_[stmt] = loop_unique_[stmt->operand];
+    }
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    if (loop_invariant_.count(stmt->lhs) > 0 &&
+        loop_invariant_.count(stmt->rhs) > 0) {
+      loop_invariant_.insert(stmt);
+    }
+    if ((loop_unique_.count(stmt->lhs) > 0 &&
+         loop_invariant_.count(stmt->rhs) > 0) &&
+        (stmt->op_type == BinaryOpType::add ||
+         stmt->op_type == BinaryOpType::sub ||
+         stmt->op_type == BinaryOpType::bit_xor)) {
+      loop_unique_[stmt] = loop_unique_[stmt->lhs];
+    }
+    if ((loop_invariant_.count(stmt->lhs) > 0 &&
+         loop_unique_.count(stmt->rhs) > 0) &&
+        (stmt->op_type == BinaryOpType::add ||
+         stmt->op_type == BinaryOpType::sub ||
+         stmt->op_type == BinaryOpType::bit_xor)) {
+      loop_unique_[stmt] = loop_unique_[stmt->rhs];
+    }
+  }
+
+  bool is_loop_unique(GlobalPtrStmt *stmt) const {
+    TI_ASSERT(num_different_loop_indices != -1);
+    std::vector<int> loop_indices;
+    loop_indices.reserve(stmt->indices.size());
+    for (auto &index : stmt->indices) {
+      auto loop_unique_index = loop_unique_.find(index);
+      if (loop_unique_index != loop_unique_.end()) {
+        if (loop_unique_index->second == -1) {
```

Comment:
Do I understand correctly that `stmt->indices` contains either a single `LoopUniqueStmt`, or one or more `LoopIndexStmt`? I.e. these two types of stmts would never mix together (call this `point-a`)?

If so, I'd prefer narrowing down the key type of `loop_uniques_` to be just `const LoopIndexStmt *`, because we know for sure that `LoopUniqueStmt` is ... unique.

1. For places where we are checking `loop_uniques_.count(stmt) > 0`, we can have a helper like this:

```cpp
// There's probably a better name...
bool is_stmt_unique(Stmt * s) const {
  if (s->is<LoopUniqueStmt>()) return true;
  if (auto *ls = s->cast<LoopIndexStmt>(); ls != nullptr)
    return loop_uniques_.count(ls) > 0;
  return false;
}
```
2. The inside this function, we just check if `indices` either contains a single `LoopUniqueStmt`, or a list of `LoopIndexStmt`s.

IMHO, this would make `point-a` more obvious, and allows us to detect if it's been violated.

Sorry I don't mean to dictate the implementation here. Let me know if this makes sense, thanks!

Reply:
> Do I understand correctly that `stmt->indices` contains either a single `LoopUniqueStmt`, or one or more `LoopIndexStmt`? I.e. these two types of stmts would never mix together (call this `point-a`)?

There's something more than this claim I want to check -- sorry for not placing it in any tests -- a `LoopUniqueStmt` add a constant should work the same as a `LoopUniqueStmt`. For example,
```python
for i, j in x:
  u = ti.loop_unique(...)
  a[u + 1] = 1 # unique because there's a u + 1
  b[u + 1, ...] = 1 # unique
  c[u, u] = 1 # unique
  d[i + 1, j + 1] = 1 # unique because there's an i and an j
  e[..., j, ..., i, ...] = 1 # unique
  f[i, i, i, i] = 1 # not unique because there's no j
```

Reply:
Ah I see, thanks!